### ISSUE TITLE
Simple IR as bytecode not readable by our or MLIR's latest tools?

### ISSUE THREAD

**Issue by dtzSiFive** [llvm/circt#5190]
Using the following IR input:
```
module {
  firrtl.circuit "ChipTop" {
    firrtl.module @ChipTop() {
      %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
    }
  }
}
```

Converting this to bytecode using `circt-as input.mlir -o input.mlirbc`, `circt-opt input.mlir -emit-bytecode -o input.mlirbc`, or `firtool -parse-only input.mlir -emit-bytecode -o input.mlirbc` produces bytecode that we cannot read back.

Additionally, the same bytecode cannot be read by recent (e.g., from our pinned LLVM) `mlir-opt --allow-unregistered-dialect` which suggests this may be an upstream issue.

The same bytecode /can/ be read by our 1.40.0 tools, including the mlir-opt from that time.
Bytecode produced similarly using above methods using 1.40.0 tools is not identical but cannot be read using our tools or recent mlir-opt.  This may suggest this is a bug in bytecode reading introduced since then (?).

When failing to read the bytecode, no diagnostic is produced or assertion encountered, the tools simply exit immediately with error code 1 ("immediately" is more obvious on the non-reduced input I first observed this with).

Here's a copy of the bytecode that was generated using `circt-as` and behaves as described (can't be read by our tools or recent mlir-opt AFAICT): [bytecode-test.tar.gz](https://github.com/llvm/circt/files/11466526/bytecode-test.tar.gz)

**Comment by dtzSiFive** [comment]
This appears to be failing due to a zero-width integer attribute, for FIRRTL this is from the `portDirections` attribute.

The following MLIR doesn't round-trip through recent `mlir-opt -emit-bytecode | mlir-opt`:

```mlir
module attributes { foo.asdf = 0 : i0 } { }
```

For some reason attribute parses as `<<NULL ATTRIBUTE>>` if you break and dump `result` here: https://github.com/llvm/llvm-project/blob/fb2c98a929aa65603e9d984307a41325e577e9d3/mlir/lib/Bytecode/Reader/BytecodeReader.cpp#L778 . (L773 on our pinned LLVM/MLIR).

This causes the attribute dictionary to take its exit path, and so on.

Older mlir-opt handles this fine.

This appears to be a new problem introduced in a recent bump, commit range: `797594a0438f2c6188e9bb3a382f041d3addcd0d..334e3776d`.

**Comment by dtzSiFive** [comment]
Reported upstream: https://github.com/llvm/llvm-project/issues/62920 .

**Comment by dtzSiFive** [comment]
Fixed upstream, fixed for us when bumping to include the change: https://github.com/llvm/circt/commit/bf2df0a84d864138e84bc5730c674403ea2bd218 .
